File: backend/internal/model/dto/notification.go
// notification.go
// 模块07 — 通知与消息：请求/响应 DTO 定义。
// 该文件对齐 docs/modules/07-通知与消息/03-API接口设计.md，覆盖收件箱、公告、定向通知、偏好、模板、内部事件和统计接口。

package dto

// InboxListReq 收件箱列表查询参数。
type InboxListReq struct {
	Page     int    `form:"page" binding:"omitempty,min=1"`
	PageSize int    `form:"page_size" binding:"omitempty,min=1,max=100"`
	Category int16  `form:"category" binding:"omitempty,oneof=1 2 3 4 5"`
	IsRead   *bool  `form:"is_read"`
	Keyword  string `form:"keyword"`
}

// NotificationInboxItem 收件箱列表项。
type NotificationInboxItem struct {
	ID           string  `json:"id"`
	Category     int16   `json:"category"`
	CategoryText string  `json:"category_text"`
	Title        string  `json:"title"`
	Content      string  `json:"content"`
	SourceModule string  `json:"source_module"`
	SourceType   *string `json:"source_type"`
	SourceID     *string `json:"source_id"`
	IsRead       bool    `json:"is_read"`
	ReadAt       *string `json:"read_at"`
	CreatedAt    string  `json:"created_at"`
}

// InboxListResp 收件箱列表响应。
type InboxListResp struct {
	List        []NotificationInboxItem `json:"list"`
	Pagination  PaginationResp          `json:"pagination"`
	UnreadCount int                     `json:"unread_count"`
}

// BatchReadNotificationsReq 批量标记已读请求。
type BatchReadNotificationsReq struct {
	IDs []string `json:"ids" binding:"required,min=1"`
}

// UnreadCountResp 未读消息计数响应。
type UnreadCountResp struct {
	Total      int                          `json:"total"`
	ByCategory NotificationUnreadByCategory `json:"by_category"`
}

// NotificationUnreadByCategory 未读消息分类统计。
// 该结构对应未读计数接口的固定分类键，避免使用无约束 map。
type NotificationUnreadByCategory struct {
	System      int `json:"system"`
	Course      int `json:"course"`
	Experiment  int `json:"experiment"`
	Competition int `json:"competition"`
	Grade       int `json:"grade"`
}

// CreateSystemAnnouncementReq 创建公告请求。
type CreateSystemAnnouncementReq struct {
	Title       string  `json:"title" binding:"required,max=200"`
	Content     string  `json:"content" binding:"required"`
	ScheduledAt *string `json:"scheduled_at"`
}

// UpdateSystemAnnouncementReq 编辑公告请求。
type UpdateSystemAnnouncementReq struct {
	Title       *string `json:"title" binding:"omitempty,max=200"`
	Content     *string `json:"content"`
	ScheduledAt *string `json:"scheduled_at"`
	IsPinned    *bool   `json:"is_pinned"`
}

// SystemAnnouncementItem 公告列表项。
type SystemAnnouncementItem struct {
	ID          string  `json:"id"`
	Title       string  `json:"title"`
	Content     string  `json:"content"`
	IsPinned    bool    `json:"is_pinned"`
	IsRead      *bool   `json:"is_read,omitempty"`
	Status      *int16  `json:"status,omitempty"`
	StatusText  *string `json:"status_text,omitempty"`
	PublishedAt *string `json:"published_at"`
	CreatedAt   *string `json:"created_at,omitempty"`
}

// SystemAnnouncementListResp 公告列表响应。
type SystemAnnouncementListResp struct {
	List       []SystemAnnouncementItem `json:"list"`
	Pagination PaginationResp           `json:"pagination"`
}

// SystemAnnouncementDetailResp 公告详情响应。
type SystemAnnouncementDetailResp struct {
	ID            string  `json:"id"`
	Title         string  `json:"title"`
	Content       string  `json:"content"`
	IsPinned      bool    `json:"is_pinned"`
	IsRead        *bool   `json:"is_read,omitempty"`
	Status        *int16  `json:"status,omitempty"`
	StatusText    *string `json:"status_text,omitempty"`
	PublishedAt   *string `json:"published_at,omitempty"`
	ScheduledAt   *string `json:"scheduled_at,omitempty"`
	UnpublishedAt *string `json:"unpublished_at,omitempty"`
	CreatedAt     *string `json:"created_at,omitempty"`
	UpdatedAt     *string `json:"updated_at,omitempty"`
}

// NotificationAnnouncementListReq 公告列表查询参数。
type NotificationAnnouncementListReq struct {
	Page     int   `form:"page" binding:"omitempty,min=1"`
	PageSize int   `form:"page_size" binding:"omitempty,min=1,max=100"`
	Status   int16 `form:"status" binding:"omitempty,oneof=1 2 3"`
}

// SendNotificationReq 发送定向通知请求。
type SendNotificationReq struct {
	Title      string `json:"title" binding:"required,max=200"`
	Content    string `json:"content" binding:"required"`
	TargetType string `json:"target_type" binding:"required"`
	TargetID   string `json:"target_id" binding:"required"`
	Category   int16  `json:"category" binding:"required,oneof=1 2 3 4 5"`
}

// NotificationPreferenceItem 通知偏好项。
type NotificationPreferenceItem struct {
	Category     int16  `json:"category"`
	CategoryText string `json:"category_text"`
	IsEnabled    bool   `json:"is_enabled"`
	IsForced     bool   `json:"is_forced"`
}

// NotificationPreferencesResp 通知偏好响应。
type NotificationPreferencesResp struct {
	Preferences []NotificationPreferenceItem `json:"preferences"`
}

// UpdateNotificationPreferencesReq 更新通知偏好请求。
type UpdateNotificationPreferencesReq struct {
	Preferences []UpdateNotificationPreferenceItem `json:"preferences" binding:"required,min=1,dive"`
}

// UpdateNotificationPreferenceItem 更新通知偏好项。
type UpdateNotificationPreferenceItem struct {
	Category  int16 `json:"category" binding:"required,oneof=1 2 3 4 5"`
	IsEnabled bool  `json:"is_enabled"`
}

// NotificationTemplateListItem 模板列表项。
type NotificationTemplateListItem struct {
	ID              string                         `json:"id"`
	EventType       string                         `json:"event_type"`
	Category        int16                          `json:"category"`
	CategoryText    string                         `json:"category_text"`
	TitleTemplate   string                         `json:"title_template"`
	ContentTemplate string                         `json:"content_template"`
	Variables       []NotificationTemplateVariable `json:"variables"`
	IsEnabled       bool                           `json:"is_enabled"`
}

// NotificationTemplateDetailResp 模板详情响应。
type NotificationTemplateDetailResp struct {
	ID              string                         `json:"id"`
	EventType       string                         `json:"event_type"`
	Category        int16                          `json:"category"`
	CategoryText    string                         `json:"category_text"`
	TitleTemplate   string                         `json:"title_template"`
	ContentTemplate string                         `json:"content_template"`
	Variables       []NotificationTemplateVariable `json:"variables"`
	IsEnabled       bool                           `json:"is_enabled"`
	CreatedAt       *string                        `json:"created_at,omitempty"`
	UpdatedAt       *string                        `json:"updated_at,omitempty"`
}

// NotificationTemplateVariable 模板变量定义。
// 该结构对应 notification_templates.variables 的固定字段。
type NotificationTemplateVariable struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Required    bool   `json:"required"`
}

// NotificationTemplateListResp 模板列表响应。
type NotificationTemplateListResp struct {
	List []NotificationTemplateListItem `json:"list"`
}

// UpdateNotificationTemplateReq 更新模板请求。
type UpdateNotificationTemplateReq struct {
	TitleTemplate   string `json:"title_template" binding:"required,max=200"`
	ContentTemplate string `json:"content_template" binding:"required"`
	IsEnabled       bool   `json:"is_enabled"`
}

// PreviewNotificationTemplateReq 预览模板请求。
type PreviewNotificationTemplateReq struct {
	Params map[string]interface{} `json:"params" binding:"required"`
}

// PreviewNotificationTemplateResp 预览模板响应。
type PreviewNotificationTemplateResp struct {
	Title   string `json:"title"`
	Content string `json:"content"`
}

// InternalSendNotificationEventReq 内部通知事件请求。
type InternalSendNotificationEventReq struct {
	EventType    string                 `json:"event_type" binding:"required"`
	ReceiverIDs  []string               `json:"receiver_ids" binding:"required,min=1"`
	Params       map[string]interface{} `json:"params" binding:"required"`
	SourceModule string                 `json:"source_module" binding:"required,max=20"`
	SourceType   string                 `json:"source_type" binding:"required,max=50"`
	SourceID     string                 `json:"source_id" binding:"required"`
}

// NotificationStatisticsReq 消息统计查询参数。
type NotificationStatisticsReq struct {
	DateFrom string `form:"date_from"`
	DateTo   string `form:"date_to"`
}

// NotificationStatisticsResp 消息统计响应。
type NotificationStatisticsResp struct {
	TotalSent  int                          `json:"total_sent"`
	TotalRead  int                          `json:"total_read"`
	ReadRate   float64                      `json:"read_rate"`
	ByCategory []NotificationCategoryStat   `json:"by_category"`
	DailyTrend []NotificationDailyTrendItem `json:"daily_trend"`
}

// NotificationCategoryStat 分类消息统计项。
type NotificationCategoryStat struct {
	Category string  `json:"category"`
	Sent     int     `json:"sent"`
	Read     int     `json:"read"`
	ReadRate float64 `json:"read_rate"`
}

// NotificationDailyTrendItem 每日消息趋势项。
type NotificationDailyTrendItem struct {
	Date string `json:"date"`
	Sent int    `json:"sent"`
	Read int    `json:"read"`
}

File: frontend/src/types/notification.ts
// notification.ts
// 模块07通知与消息类型定义：收件箱、公告、偏好、模板、统计和实时推送。

import type { ID, PaginatedData, QueryParams } from "@/types/api";

/** 通知分类：1系统 2课程 3实验 4竞赛 5成绩。 */
export type NotificationCategory = 1 | 2 | 3 | 4 | 5;

/** 公告状态：1草稿 2已发布 3已下架。 */
export type AnnouncementStatus = 1 | 2 | 3;

/** 收件箱查询参数。 */
export interface NotificationInboxListParams extends QueryParams {
  page?: number;
  page_size?: number;
  category?: NotificationCategory;
  is_read?: boolean;
  keyword?: string;
}

/** 收件箱消息项。 */
export interface NotificationInboxItem {
  id: ID;
  category: NotificationCategory;
  category_text: string;
  title: string;
  content: string;
  source_module: string;
  source_type: string | null;
  source_id: string | null;
  is_read: boolean;
  read_at: string | null;
  created_at: string;
}

/** 未读分类统计。 */
export interface NotificationUnreadByCategory {
  system: number;
  course: number;
  experiment: number;
  competition: number;
  grade: number;
}

/** 收件箱列表响应。 */
export interface NotificationInboxListResponse extends PaginatedData<NotificationInboxItem> {
  unread_count: number;
}

/** 未读计数响应。 */
export interface NotificationUnreadCount {
  total: number;
  by_category: NotificationUnreadByCategory;
}

/** 创建系统公告请求。 */
export interface CreateSystemAnnouncementRequest {
  title: string;
  content: string;
  scheduled_at?: string | null;
}

/** 更新系统公告请求。 */
export interface UpdateSystemAnnouncementRequest {
  title?: string;
  content?: string;
  scheduled_at?: string | null;
  is_pinned?: boolean;
}

/** 公告列表查询参数。 */
export interface NotificationAnnouncementListParams extends QueryParams {
  page?: number;
  page_size?: number;
  status?: AnnouncementStatus;
}

/** 系统公告项。 */
export interface SystemAnnouncementItem {
  id: ID;
  title: string;
  content: string;
  is_pinned: boolean;
  is_read?: boolean | null;
  status?: AnnouncementStatus | null;
  status_text?: string | null;
  published_at: string | null;
  created_at?: string | null;
}

/** 公告详情。 */
export interface SystemAnnouncementDetail extends SystemAnnouncementItem {
  scheduled_at?: string | null;
  unpublished_at?: string | null;
  updated_at?: string | null;
}

/** 公告列表响应。 */
export type SystemAnnouncementListResponse = PaginatedData<SystemAnnouncementItem>;

/** 发送定向通知请求。 */
export interface SendNotificationRequest {
  title: string;
  content: string;
  target_type: "all_school" | "course" | "user" | "users";
  target_id: string;
  category: NotificationCategory;
}

/** 通知偏好项。 */
export interface NotificationPreferenceItem {
  category: NotificationCategory;
  category_text: string;
  is_enabled: boolean;
  is_forced: boolean;
}

/** 通知偏好响应。 */
export interface NotificationPreferencesResponse {
  preferences: NotificationPreferenceItem[];
}

/** 更新通知偏好项。 */
export interface UpdateNotificationPreferenceItem {
  category: NotificationCategory;
  is_enabled: boolean;
}

/** 模板变量定义。 */
export interface NotificationTemplateVariable {
  name: string;
  description: string;
  required: boolean;
}

/** 模板列表项。 */
export interface NotificationTemplateListItem {
  id: ID;
  event_type: string;
  category: NotificationCategory;
  category_text: string;
  title_template: string;
  content_template: string;
  variables: NotificationTemplateVariable[];
  is_enabled: boolean;
}

/** 模板详情。 */
export interface NotificationTemplateDetail extends NotificationTemplateListItem {
  created_at?: string | null;
  updated_at?: string | null;
}

/** 模板列表响应。 */
export interface NotificationTemplateListResponse {
  list: NotificationTemplateListItem[];
}

/** 更新模板请求。 */
export interface UpdateNotificationTemplateRequest {
  title_template: string;
  content_template: string;
  is_enabled: boolean;
}

/** 模板预览响应。 */
export interface PreviewNotificationTemplateResponse {
  title: string;
  content: string;
}

/** 消息统计查询参数。 */
export interface NotificationStatisticsParams extends QueryParams {
  date_from?: string;
  date_to?: string;
}

/** 分类统计项。 */
export interface NotificationCategoryStat {
  category: string;
  sent: number;
  read: number;
  read_rate: number;
}

/** 每日消息趋势。 */
export interface NotificationDailyTrendItem {
  date: string;
  sent: number;
  read: number;
}

/** 消息统计响应。 */
export interface NotificationStatistics {
  total_sent: number;
  total_read: number;
  read_rate: number;
  by_category: NotificationCategoryStat[];
  daily_trend: NotificationDailyTrendItem[];
}

/** 通知 WebSocket 消息。 */
export interface NotificationRealtimeMessage {
  type: "unread_count_update" | "new_notification" | "new_announcement" | "pong";
  channel?: string | null;
  data?: unknown;
}
